predicate/predicate.py

```diff
@@ -60,33 +24,62 @@ def __init__(self, expr):
         self.field = None
 
     def get_field(self, instance):
-        lookup_type = 'exact' # Default lookup type
-        parts = self.lookup.split(LOOKUP_SEP)
-        num_parts = len(parts)
-        if (len(parts) > 1 and parts[-1] in QUERY_TERMS):
-            # Traverse the lookup query to distinguish related fields from
-            # lookup types.
-            lookup_model = instance
-            for counter, field_name in enumerate(parts):
+        """
+        Returns the value of this lookup on the given instance.
+
+        If the path ends with a value having the same name as a lookup
+        type (like "year", "range", etc.), the lookup must provide an
+        explicit type, such as "__exact".
+
+        Lookups can span Django fields, dicts, lists, or other integer-
+        indexed iterables.
+        """
+
+        def traverse_attr(o, attr, reraise=False):
+            # Convert numbers to integers to allow list indexing
+            if attr.isdigit():
+                attr = int(attr)
+
+            # M2M relations are a special case
+            if isinstance(o, Manager):
+                if isinstance(attr, int):
+                    return o.all()[attr]
```

Comment:
This seems likely to be pretty brittle unless the user has defined a custom ordering on the related model. What's the intended use case for integer indexing?